### ISSUE TITLE
Wan output non deterministic when --compile because of max-autotune

### ISSUE THREAD

**Issue by Sarania** [kohya-ss/musubi-tuner#268]
It's come to my attention that "max-autotune-no-cudagraphs" subtly breaks determinism when torch.compiling, at least for Wan. Outputs on the same seed with the same prompt will be 95% the same but differ in subtle details or the background. This does not seem to occur when using mode=default which actually seems faster anyway. For this reason I would suggest changing:

    parser.add_argument(
        "--compile_args",
        nargs=4,
        metavar=("BACKEND", "MODE", "DYNAMIC", "FULLGRAPH"),
        default=["inductor", "default", "False", "False"],  # <-- this second arg to default
        help="Torch.compile settings",
    )

Which I'm going to be doing on my own fork as well. I'm not sure if it affects Hunyuan but I'm going to assume it probably does. I was the one who made max-autotune-no-cudagraphs default back when I PRd that so I wanted to let you know because it took me DAYS to figure out why my refactor was producing non-deterministic results and eventually I learned it wasn't the refactor at all, it was this! So I wanted to save others the same nightmare. Cheers!

**Comment by StrongerXi** [comment]
> when using mode=default which actually seems faster anyway

Do you mean it's faster in the e2e latency, excluding initial compilation? Or just compilation is faster?

**Comment by Sarania** [comment]
In my tests it's faster e2e, I think it's basically faster up front and the same otherwise so net overall for say 832x624@81f it's about a minute faster for me. (26:00 vs 27:00)

FWIW the outputs with max-autotune are in my opinion noticeably worse, they have more structural type issues and my theory is that this might be because max-autotune is using mm_scaled/fp8 math at some point on my card(a 4070 Ti SUPER) which we know that Wan doesn't really like. That's my theory anyway. Max-autotune is VERY aggressive but it seems that doesn't really win any performance with these models and just increases comp time and subtly breaks determinism.

**Comment by sdbds** [comment]
Does max-autotune also have this issue, or is it due to no-cudagraphs? I didn't encounter any problems when testing framepack training.

**Comment by Sarania** [comment]
I haven't tested for training or for any other combination of settings, specifically I only noticed it for Wan inferences but it was quite reproducible for me in that case. It's possible it could be hardware/environment specific. I ran 8 videos in a row with default and they came out identical, 8 in a row with max-autotune-no-cudagraphs and they were all subtly different. Honestly once I realized what it was and a way to fix it, I let it go for the time being as it took me three days of brutal debugging to trace it down as I had just refactored wan_generate_video.py significantly and I thought that was the cause until I tested with native Musubi and realized it existed there too. So I was pretty stressed out and ready to move on to something else for a bit lol

**Comment by sdbds** [comment]
> I haven't tested for training or for any other combination of settings, specifically I only noticed it for Wan inferences but it was quite reproducible for me in that case. It's possible it could be hardware/environment specific. I ran 8 videos in a row with default and they came out identical, 8 in a row with max-autotune-no-cudagraphs and they were all subtly different. Honestly once I realized what it was and a way to fix it, I let it go for the time being as it took me three days of brutal debugging to trace it down as I had just refactored wan_generate_video.py significantly and I thought that was the cause until I tested with native Musubi and realized it existed there too. So I was pretty stressed out and ready to move on to something else for a bit lol
> 
> Edit: Oh FWIW these tests were under Cuda 12.8/PyTorch 2.7 with Python 3.12 on EndeavourOS linux. I would be curious to know if it happens in all environments, but the fact that it could happen at all coupled with default being faster anyway is why I decided to change it in my fork and make this issue.

This project may not be very compatible with 3.12, I think it might be better to switch back to 3.11 or 3.10.

**Comment by Sarania** [comment]
> > I haven't tested for training or for any other combination of settings, specifically I only noticed it for Wan inferences but it was quite reproducible for me in that case. It's possible it could be hardware/environment specific. I ran 8 videos in a row with default and they came out identical, 8 in a row with max-autotune-no-cudagraphs and they were all subtly different. Honestly once I realized what it was and a way to fix it, I let it go for the time being as it took me three days of brutal debugging to trace it down as I had just refactored wan_generate_video.py significantly and I thought that was the cause until I tested with native Musubi and realized it existed there too. So I was pretty stressed out and ready to move on to something else for a bit lol
> > Edit: Oh FWIW these tests were under Cuda 12.8/PyTorch 2.7 with Python 3.12 on EndeavourOS linux. I would be curious to know if it happens in all environments, but the fact that it could happen at all coupled with default being faster anyway is why I decided to change it in my fork and make this issue.
> 
> This project may not be very compatible with 3.12, I think it might be better to switch back to 3.11 or 3.10.

I've used 3.12 from the very beginning (like a week after Musubi was released) and haven't had any issues with it. I do my development in 3.12 as well, though I've maintained compat with 3.10 as well simply because that's what this repo recommends. Is there some specific thing that's known not to work in 3.12?

**Comment by sdbds** [comment]
> > > I haven't tested for training or for any other combination of settings, specifically I only noticed it for Wan inferences but it was quite reproducible for me in that case. It's possible it could be hardware/environment specific. I ran 8 videos in a row with default and they came out identical, 8 in a row with max-autotune-no-cudagraphs and they were all subtly different. Honestly once I realized what it was and a way to fix it, I let it go for the time being as it took me three days of brutal debugging to trace it down as I had just refactored wan_generate_video.py significantly and I thought that was the cause until I tested with native Musubi and realized it existed there too. So I was pretty stressed out and ready to move on to something else for a bit lol
> > > Edit: Oh FWIW these tests were under Cuda 12.8/PyTorch 2.7 with Python 3.12 on EndeavourOS linux. I would be curious to know if it happens in all environments, but the fact that it could happen at all coupled with default being faster anyway is why I decided to change it in my fork and make this issue.
> > 
> > 
> > This project may not be very compatible with 3.12, I think it might be better to switch back to 3.11 or 3.10.
> 
> I've used 3.12 from the very beginning (like a week after Musubi was released) and haven't had any issues with it. I do my development in 3.12 as well, though I've maintained compat with 3.10 as well simply because that's what this repo recommends and I want people to be able to migrate easily. Is there some specific thing that's known not to work in 3.12? 3.10 is EOL next year and 3.12 is in security status it's not like bleeding edge or anything(my system Python is actually 3.13.3 because Arch but I do literally all my Python work in 3.12.9 currently. I do have a 310 pyenv but I just haven't needed it except to test my updated requirements.txt against 3.10 again to make switching between my version and native musubi easier by using the same venv.)

This project's compatibility is the same as sd-scripts, based on 3.10, and supports up to 3.11.
I'm not sure if there are issues with 3.12, but I encountered the same problem on diffusion_pipe. That project requires 3.12...

**Comment by Sarania** [comment]
Yeah I did know that but I had 3.12 when I started and took a chance with it and never encountered issues(and I've done pretty much everything you can do with Musubi now except train framepack!) so I've just been using that. I think that's probably just what kohya-ss happens to dev and validate with so he's comfortable calling it compatible(for the same reason I actually recommend 3.12 for blissful-tuner because that's what I dev and validate in.) That said, I definitely don't know everything so that's why I asked if there was anything specific. Thanks for letting me know regardless!

Back when I first started using Musubi it wasn't even officially compatible with Linux! Honestly the worst part of Python is setting up versions/environment. I've been a programmer for about 30 years but only picked up Python about a year and a half ago. I LOVE it... EXCEPT for environments/versioning. Oof. I can handle it because I'm a tinkerer, but normal users it's rough. I'm planning to make a guided setup for this reason for my own fork like I've done before!

**Comment by Sarania** [comment]
So an update to this for anyone who saw it before, I think my initial investigation of this led me to unfortunately the wrong conclusion because it turned out to an be intermittent issue. My suspicion now is the subtle determinism breaks were caused by myself sometimes interrupting inferences with CTRL-C when my preview showed a bad gen. This would also explain the intermittent nature of the issue. You can read my comment in reply to someone here for why I originally believed it to be max-autotune:

https://github.com/kohya-ss/musubi-tuner/discussions/232#discussioncomment-13355265

I created a safe exit path for one shot mode that still cleans up and shuts down appropriately instead of exploding the model mid step and since then I haven't seen it but I am only 98% convinced because I've been tricked by this issue once before lol.
